protocol/consistencychecks.go

```diff
@@ -210,45 +223,106 @@ func (cc *ConsistencyChecks) verifySTR(str *m.SignedTreeRoot) error {
 	}
 
 	// verify hash chain
+	// TODO: clearly verify which is the client's actual expectation.
+	// See #81
 	if (str.Epoch == cc.CurrentEpoch && bytes.Equal(cc.SavedSTR, str.Signature)) ||
 		(str.Epoch == cc.CurrentEpoch+1 && str.VerifyHashChain(cc.SavedSTR)) {
 		return nil
 	}
 	return ErrorBadSTR
 }
 
-// verifyReturnedPromise verifies the returned TB
-// included in the directory's response.
-func (cc *ConsistencyChecks) verifyReturnedPromise(tb *m.TemporaryBinding,
-	uname string, key []byte,
-	str *m.SignedTreeRoot,
-	ap *m.AuthenticationPath) error {
+// verifySignedPromise checks if the returned response is match
+// with the directory's policy. Then it calls verifyFulfilledPromise
+// to verify whether issued TBs was inserted in the directory
+// as promised. Finally, it verifies the returned promise
```

Comment:
> Now the client should have enough information about that binding and can use it to verify in the next epoch. 

I'm not sure about this. I think we may be referring to two different things. What I meant by "enough information about the binding" is that a client cannot know whether the binding in the TB is indeed the binding that another client registered. So having a cached TB for Bob doesn't help Alice's client in determining whether the binding it sees for Bob is in fact a legitimate binding. Only if Alice confirms this with Bob out of band can she be sure. But we don't expect this, so my fear is that if the server equivocates about the TB, the TB verification in the next epoch will already be too late.

> If the server tries to equivocate then the client should know about it when it verifies the TB in the next epoch, right?

But you're right, having a TB is enough to detect equivocation. This is what I meant by "clients could be cautious" in my previous comment: if they want to check for equivocation they could verify a cached TB in the next epoch.